### PR TITLE
implement vnc hash types

### LINKED ISSUE #16296: VNC Capture server module's JtR output isn't parsed by JtR
The VNC capture module (`auxiliary/server/capture/vnc`) writes out John the Ripper files according to the `JOHNPWFILE` parameter. Files created by this module are not read in my JtR.

## Steps to reproduce

1. `use auxiliary/server/capture/vnc`
2. `set JOHNPWFILE /tmp/vnc_jtr`
3. `run`
4. `use auxiliary/scanner/vnc/vnc_login`
5. `set rhosts 127.0.0.1`
6. `set username user`
7. `set password p@s5w0rd`
8. `run`
9. `john /tmp/vnc_jtc_vnc`

At this point, I'd expect it to crack the hashes. Instead, it reports:

```
Using default input encoding: UTF-8
No password hashes loaded (see FAQ)
```

### Sample output

```
msf6 auxiliary(scanner/vnc/vnc_login) > run

[*] 127.0.0.1:5900        - 127.0.0.1:5900 - Starting VNC login sweep
[+] 127.0.0.1:40801 - Challenge: 00112233445566778899aabbccddeeff; Response: 6feb3cb1f07b66151656b5832341f223
[!] 127.0.0.1:5900        - No active DB -- Credential data will not be saved!
[-] 127.0.0.1:5900        - 127.0.0.1:5900 - LOGIN FAILED: :p@s5w0rd (Incorrect: Authentication failed)
[+] 127.0.0.1:37849 - Challenge: 00112233445566778899aabbccddeeff; Response: b7b9c87777661a7a2299733209bfdfce
[-] 127.0.0.1:5900        - 127.0.0.1:5900 - LOGIN FAILED: :password (Incorrect: Authentication failed)
[*] 127.0.0.1:5900        - Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
```

```
root@msfdevubuntu:~# cat /tmp/vnc_jtr_vnc 
$vnc$*00112233445566778899aabbccddeeff*6feb3cb1f07b66151656b5832341f223
$vnc$*00112233445566778899aabbccddeeff*b7b9c87777661a7a2299733209bfdfce
```

### Suggested fix

JtR Seems to expect uppercase hexadecimal. Converting the file to uppercase bears this out:

```
└─$ john --wordlist=wordlist converted_to_uppercase
Created directory: /home/kali/.john
Using default input encoding: UTF-8
Loaded 2 password hashes with 2 different salts (VNC [DES 32/64])
Press 'q' or Ctrl-C to abort, almost any other key for status
p@s5w0rd         (?)
password         (?)
2g 0:00:00:00 DONE (2022-03-07 16:18) 100.0g/s 100.0p/s 200.0c/s 200.0C/s p@s5w0rd..password
Use the "--show" option to display all of the cracked passwords reliably
Session completed
```

### PR DESCRIPTION
**Pull request by h00die**
fixes #16296

@smashery I know you had claimed the issue, but figured I could help with the hash stuff since I wrote it.  Hopefully these changes are easy to merge into the changes you've been working on.

1. removes the module writing to a john file since this is duplicate functionality already standardized and implemented in the `creds` command
2. changed the `vnc` module to not tack on `$vnc$` because that is most likely JTR specific, we'll do that when we export to jtr.
3. creates a vnc catcher for `hash_identify` so we can properly detect them, along with a lib
4. creates a `jtr` export for `vnc` hashes
5. hashcat doesn't have an *easy* cracker for `vnc` type, so I linked the latest I could find and just left it at that
6. rubocop formatting on anything I touched because its automatic
7. fixed a bug where we assumed
8.  a non-identified hash had a public (username) and when trying to export it, we crashed hard and fast.  This can be tested with 

## Verification

- [x] Follow the steps in #16296 to start the server, send a client request. (may want an easy password like `password`)
- [x] `creds -o /tmp/vnc.jtr`
- [x] `john /tmp/vnc.jtr`
- [x] **Verify** you crack the hash
- [x] `creds add hash:*00112233445566778899aabbccddeeff*6feb3cb1f07b66151656b5832341f223`
- [x] `creds -o /tmp/hashes.jtr`
- [x] **Verify** we no longer crash

### CODE REVIEW

**Comment by h00die** [comment]
```
msf6 auxiliary(server/capture/vnc) > run
[*] Auxiliary module running as background job 1.
msf6 auxiliary(server/capture/vnc) > 
[*] Started service listener on 0.0.0.0:5900 
[*] Server started.
previous
msf6 auxiliary(scanner/vnc/vnc_login) > run

[*] 127.0.0.1:5900        - 127.0.0.1:5900 - Starting VNC login sweep
[+] 127.0.0.1:37767 - Challenge: 00112233445566778899aabbccddeeff; Response: b7b9c87777661a7a2299733209bfdfce
[-] 127.0.0.1:5900        - 127.0.0.1:5900 - LOGIN FAILED: :password (Incorrect: Authentication failed)
[+] 127.0.0.1:40427 - Challenge: 00112233445566778899aabbccddeeff; Response: b7b9c87777661a7a2299733209bfdfce
[-] 127.0.0.1:5900        - 127.0.0.1:5900 - LOGIN FAILED: :password (Incorrect: Authentication failed)
[*] 127.0.0.1:5900        - Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
msf6 auxiliary(scanner/vnc/vnc_login) > creds
Credentials
===========

host       origin     service                public  private                                                             realm  private_type        JtR Format
----       ------     -------                ------  -------                                                             -----  ------------        ----------
127.0.0.1  127.0.0.1  5900/tcp (vnc_client)          *00112233445566778899aabbccddeeff*b7b9c87777661a7a2299733209bfdfce         Nonreplayable hash  vnc

msf6 auxiliary(scanner/vnc/vnc_login) > creds -o /tmp/vnc.jtr
[*] Wrote creds to /tmp/vnc.jtr
msf6 auxiliary(scanner/vnc/vnc_login) > john /tmp/vnc.jtr
[*] exec: john /tmp/vnc.jtr

Using default input encoding: UTF-8
Loaded 1 password hash (VNC [DES 32/64])
Warning: poor OpenMP scalability for this hash type, consider --fork=8
Will run 8 OpenMP threads
Proceeding with single, rules:Single
Press Ctrl-C to abort, or send SIGUSR1 to john process for status
Almost done: Processing the remaining buffered candidate passwords, if any.
Proceeding with wordlist:/usr/share/john/password.lst
password         (?)     
1g 0:00:00:00 DONE 2/3 (2022-03-13 13:27) 16.66g/s 1092Kp/s 1092Kc/s 1092KC/s 123456..Dneirfts
Use the "--show" option to display all of the cracked passwords reliably
Session completed.
```

**Comment by adfoster-r7** [comment]
@msjenkins-r7 retest this please

**Review by its0x08 — APPROVED** [review]

**Comment by adfoster-r7** [comment]
@gwillcox-r7 I just assigned @sjanusz-r7 to this, but it looks like you're already testing this functionality in a separate PR? 👀
https://github.com/rapid7/metasploit-framework/pull/16207 

Is it worth assigning this to yourself too?

**Comment by gwillcox-r7** [comment]
@adfoster-r7 Happy to pick this up seeing as the two are related 👍 Otherwise no issues with @sjanusz-r7 picking this up if he is interested

File: lib/metasploit/framework/password_crackers/jtr/formatter.rb [hunk 1]
```diff
@@ -67,6 +67,9 @@ def hash_to_jtr(cred)
       # with a sample hash of b31d032cfdcf47a399990a71e43c5d2a:144816. So this just outputs
       # The hash as *hash*: so that it is both JTR and hashcat compatible
       return "#{cred.private.data}:"
+    when /vnc/
+      # add a beggining * if one is missing
```
Comment (gwillcox-r7) [inline lib/metasploit/framework/password_crackers/jtr/formatter.rb:71]:
```suggestion
      # add a beginning * if one is missing
```

File: lib/metasploit/framework/password_crackers/jtr/formatter.rb [hunk 2]
```diff
@@ -76,7 +79,9 @@ def hash_to_jtr(cred)
       # /ssha/
       # /raw-sha512/
       # /raw-sha256/
-      return "#{cred.public.username}:#{cred.private.data}:#{cred.id}:"
+      # this also handles *other* and it isn't guaranteed to have a public
```
Comment (gwillcox-r7) [inline lib/metasploit/framework/password_crackers/jtr/formatter.rb:82]: Need to be more specific here, this assumes too much background knowledge from the reader and the terms "other" and "it isn't" aren't clear here.

Comment (gwillcox-r7) [reply lib/metasploit/framework/password_crackers/jtr/formatter.rb:82]: I changed the wording here to `This also handles *other* type credentials which aren't guaranteed to have a public` but let me know if you think this needs changing.

Comment (gwillcox-r7) [reply lib/metasploit/framework/password_crackers/jtr/formatter.rb]: Addressed with https://github.com/rapid7/metasploit-framework/pull/16341/commits/1408ffbab7d8fa24fa0a6fd4ff6abcb7ca5dce04

File: lib/metasploit/framework/password_crackers/jtr/formatter.rb (hunk 1 above)
Comment (gwillcox-r7) [reply lib/metasploit/framework/password_crackers/jtr/formatter.rb]: Fixed in https://github.com/rapid7/metasploit-framework/pull/16341/commits/1408ffbab7d8fa24fa0a6fd4ff6abcb7ca5dce04

**Comment by gwillcox-r7** [comment]
Verified can capture and crack the hash:

```
msf6 auxiliary(scanner/vnc/vnc_login) > creds
Credentials
===========

host  origin  service  public  private  realm  private_type  JtR Format
----  ------  -------  ------  -------  -----  ------------  ----------

msf6 auxiliary(scanner/vnc/vnc_login) > run

[*] 127.0.0.1:5900        - 127.0.0.1:5900 - Starting VNC login sweep
[+] 127.0.0.1:38299 - Challenge: 00112233445566778899aabbccddeeff; Response: b7b9c87777661a7a2299733209bfdfce
[-] 127.0.0.1:5900        - 127.0.0.1:5900 - LOGIN FAILED: :password (Incorrect: Authentication failed)
[+] 127.0.0.1:43993 - Challenge: 00112233445566778899aabbccddeeff; Response: b7b9c87777661a7a2299733209bfdfce
[-] 127.0.0.1:5900        - 127.0.0.1:5900 - LOGIN FAILED: :password (Incorrect: Authentication failed)
[*] 127.0.0.1:5900        - Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
msf6 auxiliary(scanner/vnc/vnc_login) > show options

Module options (auxiliary/scanner/vnc/vnc_login):

   Name              Current Setting                     Required  Description
   ----              ---------------                     --------  -----------
   BLANK_PASSWORDS   false                               no        Try blank passwords for all users
   BRUTEFORCE_SPEED  5                                   yes       How fast to bruteforce, from 0 to 5
   DB_ALL_CREDS      false                               no        Try each user/password couple stored in the current database
   DB_ALL_PASS       false                               no        Add all passwords in the current database to the list
   DB_ALL_USERS      false                               no        Add all users in the current database to the list
   DB_SKIP_EXISTING  none                                no        Skip existing credentials stored in the current database (Acc
                                                                   epted: none, user, user&realm)
   PASSWORD          password                            no        The password to test
   PASS_FILE         /home/gwillcox/git/metasploit-fram  no        File containing passwords, one per line
                     ework/data/wordlists/vnc_passwords
                     .txt
   Proxies                                               no        A proxy chain of format type:host:port[,type:host:port][...]
   RHOSTS            127.0.0.1                           yes       The target host(s), see https://github.com/rapid7/metasploit-
                                                                   framework/wiki/Using-Metasploit
   RPORT             5900                                yes       The target port (TCP)
   STOP_ON_SUCCESS   false                               yes       Stop guessing when a credential works for a host
   THREADS           1                                   yes       The number of concurrent threads (max one per host)
   USERNAME          user                                no        A specific username to authenticate as
   USERPASS_FILE                                         no        File containing users and passwords separated by space, one p
                                                                   air per line
   USER_AS_PASS      false                               no        Try the username as the password for all users
   USER_FILE                                             no        File containing usernames, one per line
   VERBOSE           true                                yes       Whether to print output for all attempts

msf6 auxiliary(scanner/vnc/vnc_login) > creds
Credentials
===========

host       origin     service                public  private                                                             realm  private_type        JtR Format
----       ------     -------                ------  -------                                                             -----  ------------        ----------
127.0.0.1  127.0.0.1  5900/tcp (vnc_client)          *00112233445566778899aabbccddeeff*b7b9c87777661a7a2299733209bfdfce         Nonreplayable hash  vnc

msf6 auxiliary(scanner/vnc/vnc_login) > creds -o /tmp/vnc.jtr
[*] Wrote creds to /tmp/vnc.jtr
msf6 auxiliary(scanner/vnc/vnc_login) >
```

And cracking it:

```
 ~/john-1.9.0-Jumbo-1/run  john /tmp/vnc.jtr                                                                      ✔ │ 3.0.2 Ruby 
Using default input encoding: UTF-8
Loaded 1 password hash (VNC [DES 32/64])
Warning: poor OpenMP scalability for this hash type, consider --fork=2
Will run 2 OpenMP threads
Proceeding with single, rules:Single
Press 'q' or Ctrl-C to abort, almost any other key for status
Almost done: Processing the remaining buffered candidate passwords, if any.
Proceeding with wordlist:password.lst, rules:Wordlist
password         (?)
1g 0:00:00:00 DONE 2/3 (2022-03-22 16:51) 20.00g/s 327680p/s 327680c/s 327680C/s 123456..betabeta
Use the "--show" option to display all of the cracked passwords reliably
Session completed
 ~/john-1.9.0-Jumbo-1/run           
```

**Comment by gwillcox-r7** [comment]
And looks like we no longer crash:

```
msf6 auxiliary(scanner/vnc/vnc_login) > creds add hash:*00112233445566778899aabbccddeeff*6feb3cb1f07b66151656b5832341f223
msf6 auxiliary(scanner/vnc/vnc_login) > creds -o /tmp/vnc.jtr
[*] Wrote creds to /tmp/vnc.jtr
msf6 auxiliary(scanner/vnc/vnc_login) > creds
Credentials
===========

host       origin     service                public  private                                                             realm  private_type        JtR Format
----       ------     -------                ------  -------                                                             -----  ------------        ----------
                                                     *00112233445566778899aabbccddeeff*6feb3cb1f07b66151656b5832341f223         Nonreplayable hash  
127.0.0.1  127.0.0.1  5900/tcp (vnc_client)          *00112233445566778899aabbccddeeff*b7b9c87777661a7a2299733209bfdfce         Nonreplayable hash  vnc

msf6 auxiliary(scanner/vnc/vnc_login) > 
```

**Comment by gwillcox-r7** [comment]
LGTM will land this now 👍

**Review by gwillcox-r7 — APPROVED** [review]

**Comment by gwillcox-r7** [comment]
# Release Notes
Support for capturing and exporting VNC hashes for cracking via John The Ripper has now been added into Metasploit.

**Comment by gwillcox-r7** [comment]
# Release Notes
Fixed a bug where the `auxiliary/server/capture/vnc` module would not output hashes in a format compatible with John The Ripper and also fixed a bug that was causing crashes due to assuming hashes always had an associated username. Support has also been added for exporting VNC hashes into a JTR compatible format for later cracking and the `hash_identify` function has been updated to properly identify VNC hashes allowing for better hash detection.

**Comment by gwillcox-r7** [comment]
Heads up for whoever is doing wrapup this week, this is technically both a bug fix and an enhancement merged into one PR hence the dual labels here.

**Comment by adfoster-r7** [comment]
@gwillcox-r7 I believe the wrapup generator will preference placing this in the `rn-fix` section instead of the `rn-enhancement`, that's a somewhat arbitrary decision - but just a headsup

**Pull request merged**